src/Utilities/SingleStackUtils/SingleStackUtils.jl

```diff
@@ -453,4 +454,258 @@ function dict_of_nodal_states(
+        # to the cartesian space `Q[i, var, j]`
+        if interp
+            cart_ind = CartesianIndices((Nk, size(prognostic, 3)))
+        else
+            cart_ind = CartesianIndices((Nqk, size(prognostic, 3)))
+        end
+        args = (bl, grid, states, vrange, i, j, cart_ind, interp)
+        BL, G, S, VR, TI, TJ, CI, IN = typeof.(args)
+        if interp
+            len = size(prognostic, 3) * Nk + 1
+        else
+            len = size(prognostic, 3) * Nqk
+        end
+        new{len, BL, G, S, VR, TI, TJ, CI, IN}(args...)
+    end
+end
+
+# Convenience wrapper
+GridStack(solver_config; kwargs...) = GridStack(
+    solver_config.dg.balance_law,
+    solver_config.dg.grid,
+    solver_config.Q,
+    solver_config.dg.state_auxiliary,
+    solver_config.dg.state_gradient_flux,
+    solver_config.dg.states_higher_order[2];
+    kwargs...,
+)
+
+Base.length(gs::GridStack{N}) where {N} = N
+
+# Helper function
+get_state(v, state, vid⁻, vid⁺, ev⁻, ev⁺, J⁻, J⁺) =
+    (J⁻ * state[vid⁻, v, ev⁻] + J⁺ * state[vid⁺, v, ev⁺]) / (J⁻ + J⁺)
+
+to_cpu(state) =
+    array_device(state) isa CPU ? state.realdata : Array(state.realdata)
+
+function interp_top(state, args, n_vars, bl, st)
+    vs = Vars{vars_state(bl, st, eltype(state))}
+    if n_vars ≠ 0
+        return vs(map(v -> get_state(v, state, args...), 1:n_vars))
+    else
+        return nothing
+    end
+end
+
+function interp_bot(state, vid⁻, ev⁻, n_vars, bl, st)
+    vs = Vars{vars_state(bl, st, eltype(state))}
+    if n_vars ≠ 0
+        return vs(map(v -> state[vid⁻, v, ev⁻], 1:n_vars))
+    else
+        return nothing
+    end
+end
+
+function no_interp(state, ijk, ev, n_vars, bl, st)
+    vs = Vars{vars_state(bl, st, eltype(state))}
+    if n_vars ≠ 0
+        return vs(map(v -> state[ijk, v, ev], 1:n_vars))
+    else
+        return nothing
+    end
+end
+
+function Base.iterate(gs::GridStack, iter_state = 1)
+    iter_state > length(gs) && return nothing
+
+    # extract grid information
+    grid = gs.grid
+    FT = eltype(grid)
+    N = polynomialorders(grid)
+    dim = dimensionality(grid)
+    Nq = N .+ 1
+    # Code assumes the same polynomial order in all horizontal directions
+    @inbounds begin
+        Nq1 = Nq[1]
+        Nq2 = Nq[2]
+        Nqk = dim == 2 ? 1 : Nq[dim]
+    end
+    Np = dofs_per_element(grid)
+    states = gs.states
+    bl = gs.bl
+    interp = gs.interp
+
+    # bring `Q` to the host if needed
+
+    prognostic = to_cpu(states.prog)
+    auxiliary = to_cpu(states.aux)
+    diffusive = to_cpu(states.diffusive)
+    hyperdiffusive = to_cpu(states.hyperdiffusive)
+
+    n_vars_prog = size(prognostic, 2)
+    n_vars_aux = size(auxiliary, 2)
+    n_vars_diff = size(diffusive, 2)
+    n_vars_hd = size(hyperdiffusive, 2)
+
+    elemtobndy = convert(Array, grid.elemtobndy)
+    vmap⁻ = convert(Array, grid.vmap⁻)
+    vmap⁺ = convert(Array, grid.vmap⁺)
+    vgeo = convert(Array, grid.vgeo)
```

Comment:
This makes this code very tightly bound to the internals of grids. It would be better to add interface functions in that module to support the functionality you need below than to do this.

Reply:
> This makes this code very tightly bound to the internals of grids.
Meaning the use of `elemtobndy`,`vmap⁻`,`vmap⁺`,`vgeo`? Or the `convert`?

> It would be better to add interface functions in that module to support the functionality you need below than to do this.

Providing an interface to access elements of `elemtobndy`,`vmap⁻`,`vmap⁺`,`vgeo`?

Reply:
Yes, the use of `elemtobndy` etc. And the interface should be to find out what you want to find out rather than to access the data structures to get the information. But as you say below, this provides a concrete data point for what's desired.